### ISSUE TITLE
LOGFILE command line option or configuration statement 

### ISSUE THREAD

**Issue by arfineman** [SDL-Hercules-390/hyperion#354]
It would be nice to have a command line option or a configuration statement for Hercules to save the log file. This feature is currently available as an output redirect option, but requires specific method of invocation.

**Comment by Fish-Git** [comment]
This request seems quite reasonable to me and I'm actually surprised that no Hercules developer (or user for that matter) has ever thought, until now, to provide such an option! It's a good one IMHO.

Unless no other developer objects or someone comes up with a reason why we _shouldn't_ do this, I'll try to get this done right away! It certainly _seems_ noncontroversial and simple enough to do!

Of course, "right away" == "soon" == "whenever I get a round tuit" == "whenever I find the time to do it" == "at some unspecified time in the future" == "maybe tomorrow, maybe next week, maybe next year, etc." &nbsp; `;-)`

**Comment by ivan-w** [comment]
`log <filename>` isn't good enough ? (already exists)

**Comment by arfineman** [comment]
The LOG command in a rc file can partially accomplish this. But all startup messages would be lost.
Best regards,

**Comment by Fish-Git** [comment]
Ivan,

Aaron is correct. The Hercules `log` command is not available until _after_ Hercules "comes up", thereby causing the logfile to be missing many important startup/initialization messages (such as what version is being executed, what build options were used, etc.), all of which are important to have for problem analysis purposes.

As an example, here's the result of my adding a "log" statement at the very, very beginning (i.e. as the very first statement) of Hercules's sample/example configuration file, `hercules.cnf`, that comes delivered in our repository:

```
13:45:08 HHC17012I MSGLEVEL = verbose nodebug emsgloc
13:45:08 HHC02204I CPUSERIAL      set to 002623
13:45:08 HHC02204I CPUMODEL       set to 3090
13:45:08 HHC02204I MODEL          set to hardware(EMULATOR) capacity(EMULATOR) perm() temp()
13:45:08 HHC02204I PLANT          set to ZZ
13:45:08 HHC02204I MANUFACTURER   set to HRC
13:45:08 HHC02204I LPARNAME       set to HERCULES
13:45:08 HHC02204I CPUVERID       set to FD
13:45:08 HHC17003I MAIN     storage is 64M (mainsize); storage is not locked
13:45:08 HHC17003I EXPANDED storage is 0 (xpndsize); storage is not locked
13:45:08 HHC02204I NUMCPU         set to 1
13:45:08 HHC02204I MAXCPU         set to 8
13:45:08 HHC02204I ARCHLVL        set to z/Arch
13:45:08 HHC00898W Facility( 006_ASN_LX_REUSE ) *Disabled for z/Arch
13:45:08 HHC00007I Previous message from function 'facility_enable_disable' at facility.c(3621)
13:45:08 HHC02204I LOADPARM       set to 0120....
13:45:08 HHC02204I TZOFFSET       set to +0200
13:45:08 HHC01474I Using internal codepage conversion table default
13:45:08 HHC02204I SHCMDOPT       set to DISABLE  NODIAG8
13:45:08 HHC02204I DIAG8CMD       set to DISABLE  NOECHO
13:45:08 HHC02256W Command 'PANRATE' is deprecated; use PANOPT RATE=xxx instead
13:45:08 HHC00007I Previous message from function 'panrate_cmd' at hsccmd.c(4131)
13:45:08 HHC02204I PANRATE        set to FAST
13:45:08 HHC01250E 0:000C Card: error in function access(): No such file or directory
13:45:08 HHC00007I Previous message from function 'cardrdr_init_handler' at cardrdr.c(322)
13:45:08 HHC01463E 0:000C device initialization failed
13:45:08 HHC00007I Previous message from function 'attach_device' at config.c(1330)
13:45:08 HHC00100I Thread id 00001c7c, prio 4, name 'console_connect' started
13:45:08 HHC01024I Waiting for console connections on port 3270
13:45:13 HHC01603I *
13:45:14 HHC01603I *
13:45:15 HHC01603I *
13:45:22 HHC01603I * Ivan test...
13:45:24 HHC01603I *
13:45:25 HHC01603I *
13:45:25 HHC01603I *
13:45:27 HHC01603I exit
13:45:27 HHC01420I Begin Hercules shutdown
13:45:27 HHC01423I Calling termination routines
13:45:27 HHC00101I Thread id 00001a30, prio 2, name 'Processor CP00' ended
13:45:27 HHC00101I Thread id 00001c7c, prio 4, name 'console_connect' ended
13:45:27 HHC01427I Main storage released
13:45:27 HHC01427I Expanded storage released
13:45:27 HHC01422I Configuration released
13:45:27 HHC00101I Thread id 000012e0, prio 7, name 'timer_thread' ended
13:45:27 HHC00101I Thread id 00000970, prio 5, name 'panel_display' ended
```

Notice it's missing the version messages and many other important problem determination messages.

We need a new command-line option that allows the user to specify the name of the logfile _without_ having to use "command-line redirection syntax", that ensures the logfile will not only always be created (which might not occur if the user uses incorrect syntax on their command-line), but will also always be _complete_ (i.e. always contain _ALL_ startup/initialization messages).

That's all Aaron is asking for.

I'm thinking maybe `-o / --output / --logfile`.

**Comment by wrljet** [comment]
One consideration is: this sort of feature goes against the basic philosophy of *nix.

If you want to redirect output to a file, you don't do that with special features of an application program.
*nix provides mechanisms to do that that work with every user-space program.

Bill

**Comment by Fish-Git** [comment]
Understood, Bill, but the request is not to change that ability (today's ability to redirect output to a file will remain), but rather to simply provide an additional way to accomplish the same thing.

IMO, requesting that a logfile be produced _should_ be controlled via a command-line option rather than by redirection given how Hercules behaves/is designed. It is, after all, not your typical command-line program that simply performs a query and generates text output as its response!

Most command-line programs, when their output (stdout) is redirected to a file, don't write anything to the screen (except MAYBE a few error messages). Hercules behaves quite differently: when you redirect it's output to a file, it _STILL_ writes output to the screen (and they're not just error/warning messages either). It's almost like a daemon. Quite unlike a typical enter a command and get a text response type program.

IMO a new command-line option seems entirely reasonable to me.

But of course my opinion is my own and doesn't carry any more weight than anyone else's. If the other Hercules developers also feel Aaron's suggested enhancement shouldn't be made, then it won't be made.

**Comment by wrljet** [comment]
You're right.  I agree.

**Comment by Fish-Git** [comment]
Fixed by commit ac95c4203dc30010a6c304392368fe24e8de5907.

Closing.
